src/Package.zig

```diff
@@ -1,8 +1,186 @@
+        var buf: std.ArrayListUnmanaged(u8) = .initBuffer(&result.bytes);
+        buf.appendSliceAssumeCapacity(name[0..@min(name.len, 32)]);
+        buf.appendAssumeCapacity('-');
+        buf.appendSliceAssumeCapacity(ver[0..@min(ver.len, 32)]);
+        buf.appendAssumeCapacity('-');
+        var sizedhash: [9]u8 = undefined;
+        std.mem.writeInt(u32, sizedhash[0..4], size, .little);
+        sizedhash[4..].* = digest[0..5].*;
+        _ = std.base64.url_safe_no_pad.Encoder.encode(buf.addManyAsArrayAssumeCapacity(12), &sizedhash);
+        @memset(buf.unusedCapacitySlice(), 0);
+        return result;
+    }
+
+    /// Produces "$hashiname-N-$sizedhash". For packages that lack "build.zig.zon" metadata.
+    /// * hashiname is [5..][0..24] bytes of the SHA-256, urlsafe-base64-encoded, for a total of 32 bytes encoded
+    /// * the semver section is replaced with a hardcoded N which stands for
+    ///   "naked". It acts as a version number so that any future updates to the
+    ///   hash format can tell this hash format apart. Note that "N" is an
+    ///   invalid semver.
+    /// * sizedhash is the same as in `init`.
+    ///
+    /// The hash is broken up this way so that "sizedhash" can be calculated
+    /// exactly the same way in both cases, and so that "name" and "hashiname" can
+    /// be used interchangeably in both cases.
+    pub fn initNaked(digest: Digest, size: u32) Hash {
+        var name: [32]u8 = undefined;
+        _ = std.base64.url_safe_no_pad.Encoder.encode(&name, digest[5..][0..24]);
+        return init(digest, &name, "N", size);
+    }
+
+    /// Produces a unique hash based on the path provided. The result should
+    /// not be user-visible.
+    pub fn initPath(sub_path: []const u8, is_global: bool) Hash {
+        var result: Hash = .{ .bytes = @splat(0) };
+        var i: usize = 0;
+        if (is_global) {
+            result.bytes[0] = '/';
+            i += 1;
+        }
+        if (i + sub_path.len <= result.bytes.len) {
+            @memcpy(result.bytes[i..][0..sub_path.len], sub_path);
+            return result;
+        }
+        var bin_digest: [Algo.digest_length]u8 = undefined;
+        Algo.hash(sub_path, &bin_digest, .{});
+        _ = std.fmt.bufPrint(result.bytes[i..], "{}", .{std.fmt.fmtSliceHexLower(&bin_digest)}) catch unreachable;
+        return result;
+    }
+};
+
+pub const MultihashFunction = enum(u16) {
+    identity = 0x00,
+    sha1 = 0x11,
+    @"sha2-256" = 0x12,
+    @"sha2-512" = 0x13,
+    @"sha3-512" = 0x14,
+    @"sha3-384" = 0x15,
+    @"sha3-256" = 0x16,
+    @"sha3-224" = 0x17,
+    @"sha2-384" = 0x20,
+    @"sha2-256-trunc254-padded" = 0x1012,
+    @"sha2-224" = 0x1013,
+    @"sha2-512-224" = 0x1014,
+    @"sha2-512-256" = 0x1015,
+    @"blake2b-256" = 0xb220,
+    _,
+};
+
+pub const multihash_function: MultihashFunction = switch (Hash.Algo) {
+    std.crypto.hash.sha2.Sha256 => .@"sha2-256",
+    else => @compileError("unreachable"),
```

Comment:
just out of curiosity, why is `@compileError()` better than plain `unreachable` here given this is comptime anyway?

Reply:
This code is moved, it's not edited. That said I'm also the original author here, so I suppose I owe an explanation. Here it is: :shrug:

I'll go ahead and toss that change in